astropy/modeling/core.py

```diff
@@ -2747,6 +2791,52 @@ def __str__(self):
         ]
         return super()._format_str(keywords=keywords)
 
+    def _generate_input_output_units_dict(self, mapping, attr):
+        """
+        This method is used to transform dict or bool settings from
+        submodels into a single dictionary for the composite model,
+        taking into account renaming of input parameters.
+        """
+        d = {}
+        for inp, (model, orig_inp) in mapping.items():
+            mattr = getattr(model, attr)
+            if isinstance(mattr, dict):
+                if orig_inp in mattr:
+                    d[inp] = mattr[orig_inp]
+            elif isinstance(mattr, bool):
+                d[inp] = mattr
+
+        if d:  # Note that if d is empty, we just return None
+            return d
+
+    @property
+    def _supports_unit_fitting(self):
+        return False
+
+    @property
+    def input_units_allow_dimensionless(self):
+        return self._generate_input_output_units_dict(self._inputs_map,
+                                                      'input_units_allow_dimensionless')
+
+    @property
+    def input_units_strict(self):
+        return self._generate_input_output_units_dict(self._inputs_map,
+                                                      'input_units_strict')
```

Comment:
For the following case, is the output expected?

```
g = Gaussian1D(10 * u.Jy, 2 * u.m, 1 * u.m)
offx = Shift(1)
m = g & offx
# If one of the models has `input_units` should it be required that all models define units?
m.input_units
Out[59]: {'x0': Unit("m")}

m.input_units_strict
Out[58]: {'x0': False, 'x1': True}

m(1 * u.m, 2)
Out[61]: (<Quantity 6.065306597126334 Jy>, 3.0)
```

Reply:
To expand a bit on this, I would expect only this to work:
```
g = Gaussian1D(10 * u.Jy, 2 * u.m, 1 * u.m)
offx = Shift(1 * u.m)
m = g & offx
m(1*u.m,2*u.m)
Out[69]: (<Quantity 6.065306597126334 Jy>, <Quantity 3.0 m>)

```

Reply:
Just to check, what is wrong with the first example? The fact that ``m.input_units_strict`` is set for ``x1`` doesn't matter as there are no ``input_units`` defined for ``x1``. Is there any reason to disallow that case? (I mean we could certainly include a warning that says that  ``m.input_units_strict['x1']`` will get ignored.

Reply:
@astrofrog I'm not concerned with `input_units_strict`. I guess I'm not sure we should allow a mixture of unitless and "with_units" models. Could one get a wrong answer because this is allowed?